Keras deployer

eavidan opened the pull request:
Based on the tensorflow deployer, I have implemented this Keras deployer
Currentlly a docker build command needs to be run to create the keras-container image
```
docker build -f dockerfiles/KerasDockerfile -t keras-container .
```
This is until build_docker_images.sh will be updated with the build commands

a working example would be:
```
from clipper_admin import ClipperConnection, DockerContainerManager
from clipper_admin.deployers import keras as keras_deployer
import keras

# run this first
# docker build -f dockerfiles/KerasDockerfile -t keras-container .

# creating a simple Keras model
inpt = keras.layers.Input(shape=(1,))
out = keras.layers.multiply([inpt, inpt])
model = keras.models.Model(inputs=inpt, outputs=out)

clipper_conn = ClipperConnection(DockerContainerManager())

# Connect to an already-running Clipper cluster
clipper_conn.connect()

def predict(model, inputs):
    return [model.predict(x) for x in inputs]

keras_deployer.deploy_keras_model(clipper_conn=clipper_conn, 
    name="pow", 
    version="1",
    input_type="ints", 
    func=predict, 
    model_path_or_object=model, 
    base_image='keras-container')

# sending an inference request
import requests
import json

req_json = json.dumps({
    "input": [1, 2, 4, 6]
})

headers = {"Content-type": "application/json"}
response = requests.post("http://localhost:1337/keras-pow/predict", headers=headers,
    data=req_json)
```

eavidan committed:
Keras deployer
currently need to run the docker build before using the deployer
`docker build -f dockerfiles/KerasDockerfile -t keras-container .`
The build commands will be added to the build_docker_images.sh

eavidan committed:
example file removed

AmplabJenkins commented:
Can one of the admins verify this patch?

simon-mo commented:
This is awesome! Can you add a test like https://github.com/ucbrise/clipper/blob/develop/integration-tests/deploy_tensorflow_models.py?

eavidan committed:
- integration test add for keras deployer
- exception is raised when model_path_or_object passed to keras deployer is neither a Model object or an *.h5 file

eavidan commented:
just pushed an integration test and a small exception for the model validation part
should I add the following lines to `build_docker_images`?
```
create_image keras-container KerasDockerfile $public py &
create_image keras35-container KerasDockerfile $public py35 &
create_image keras36-container KerasDockerfile $public py36 &
```

simon-mo commented:
Thanks again! Sorry about the late reply. I can take care of that. We are in the midst of integration testing infrastructure refactoring (https://github.com/ucbrise/clipper/pull/595). I will add building container support to this PR once CI is ready.

ivrt commented:
Hi, I'm trying to run this locally and I am running into an issue running the following command:

`docker build -f dockerfiles/KerasDockerfile -t keras-container .`

The error message I'm getting is:

`pull access denied for clipper-base, repository does not exist or may require 'docker login'`

It seems like there is no such image as clipper-base on docker hub. Not super familiar with docker but any ideas what I'm doing wrongly?

eavidan commented:
my bad
I believe I built it locally from Py2RPCDockerfile
```
docker build -f dockerfiles/Py2RPCDockerfile -t clipper-base .
```

@simon-mo, sorry about that. should I update the docker file with `FROM ${REGISTRY}/${RPC_VERSION}-rpc:${CODE_VERSION}` ?

simon-mo commented:
Yes please that would be great! In the future, building clipper images will be simpler. But for now this is the correct approach.

eavidan committed:
Keras dockerfile base image will be taken from dockerhub

harmw commented:
Just build a model container using this PR, awesome! One problem we ran into was the `keras-container` expects to find the `.h5` model as a file on the local filesystem, but we packaged it inside our pickled function (`func.pkl`). Would it make sense having the following added to this PR?

```
diff --git a/containers/python/keras_container.py b/containers/python/keras_container.py
index e2dd47f..cb96bc9 100644
--- a/containers/python/keras_container.py
+++ b/containers/python/keras_container.py
@@ -27,7 +27,12 @@ class KerasContainer(rpc.ModelContainerBase):
             dir=path, predict_fname=predict_fname)
         self.predict_func = load_predict_func(predict_path)

-        self.model = load_model(os.path.join(path, "keras_model.h5"))
+        model_file = os.path.join(path, "keras_model.h5")
+        if os.path.exists(model_file):
+            self.model = load_model(model_file)
+        else:
+            print("keras_model.h5 not found, it's probably pickled")
+            self.model = None

     def predict_ints(self, inputs):
         preds = self.predict_func(self.model, inputs)
```

One downside of using the stock deployers is that they build&ship in one step, something we split in a two-stage process. Our pickle step is as simple as just `save_python_function(None, make_prediction)`.

rkooo567 commented:
@simon-mo  Will we merge this PR?

withsmilo commented:
@simon-mo @rkooo567 
How about try to merge this PR? I think that the Keras deployer is too awesome and many people wait for this.

simon-mo commented:
We are still missing
1. Keras integration test
2. Keras docker container 

@rkooo567 do you want to handle that?

rkooo567 commented:
@simon-mo Sure. I can work on it from 18th. 

Btw, I can see the integration test as well as KerasContainer + Dockerfile. Are there any missing parts from those two?

simon-mo commented:
sorry I mean we need to add them to the CI. So add a line here:
https://github.com/ucbrise/clipper/blob/8051a4e5ecf814177c0442b0b56760e4e505beb1/bin/shipyard/clipper_docker.cfg.py#L159-L170

and here:
https://github.com/ucbrise/clipper/blob/8051a4e5ecf814177c0442b0b56760e4e505beb1/bin/shipyard/clipper_test.cfg.py#L15-L29

rkooo567 commented:
@simon-mo Gotcha! That sounds simple. Let's wait for @eavidan, and see if he can add one more commit first. If he wouldn't appear, I will add them within a few days.

rkooo567 committed:
Merge branch 'develop' into keras_deployer

rkooo567 committed:
keras_deployer: added code for CI testing. The test fails locally. We should see if our CI can build dockerfile.

rkooo567 commented:
Jenkins test this please

rkooo567 commented:
Jenkins test this please

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1958/
Test FAILed.

simon-mo committed:
Rename Keras Dockerfile

rkooo567 commented:
Jenkins test this please

rkooo567 commented:
@simon-mo This also failed with mxnet test. Maybe it is better reverting mxnet update for now? I think it is pretty unstable

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1963/
Test FAILed.

simon-mo committed:
Merge branch 'develop' into keras_deployer

simon-mo commented:
@rkooo567 i don't think this failed with mxnet yet. the rbac one failed with k8s but other tests like tensorflow and sparkml also failed.

withsmilo commented on clipper_admin/clipper_admin/deployers/keras.py:
```diff
@@ -0,0 +1,243 @@
+        import keras
+
+        # run this first
+        # docker build -f dockerfiles/KerasDockerfile -t keras-container .
```
This comment is not proper here.

rkooo567 replied:
Let me remove it

withsmilo commented on dockerfiles/KerasContainerDockerfile:
```diff
@@ -0,0 +1,14 @@
+
+RUN pip install -q keras
+RUN pip install -q tensorflow
+RUN pip install -q numpy==1.16 # MXNet requires numpy < 1.15
```
MXNet requires `numpy<=1.15.2,>=1.8.2`.
https://github.com/apache/incubator-mxnet/blob/master/python/setup.py#L33

rkooo567 replied:
I will remove the comment. I guess mxnet 1.4.0 requires <1.15 (because I saw from my pip). This part will keep changing as mxnet 1.4.* is upgraded.

withsmilo replied:
OK. That comment is a little confused for me.

rkooo567 commented:
@withsmilo @simon-mo I will use tensorflow=1.16.* for `KerasContainerDockerfile`, which is same as the one defined in `PythonDevDockerfile`.

rkooo567 committed:
Fixed unnecessary comment, changed tensorflow version

rkooo567 committed:
Merge branch 'keras_deployer' of https://github.com/eavidan/clipper into keras_deployer

rkooo567 commented:
Jenkins test this please

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1969/
Test FAILed.

rkooo567 commented:
@simon-mo. 

Should we push a KerasContainerDockerfile to our Dockerhub first? Seems like we start CI from prepull (to utilize cache), and it pulls the image from our hub. I suspect it is broken there because our KerasContainerDockerfile:develop is not in Clipper Dockerhub

simon-mo commented:
@rkooo567 no because the prepull process will ignore any errors. Its just there to speed up caching. The real issue is:
``` [keras-container] Invalid requirement: 'tensorflow=1.6.*' 
 [keras-container] = is not a valid operator. Did you mean == ? 
``` 

I'm fixing it now

simon-mo committed:
Fix install ==

rkooo567 commented:
Jenkins test this please

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1972/
Test FAILed.

rkooo567 commented:
I found this.
```
Starting MXNetContainer container
ERROR: CLIPPER_MODEL_NAME environment variable must be set
Encountered error not related to missing packages. Please refer to the container log to diagnose.
```

rkooo567 commented:
Looks like it doesn't even run mxnet container. 

```
===== start: integration_py2_mxnet =====

[integration_py2_mxnet] Starting Trial 0 with timeout 2400.0 seconds

[integration_py2_mxnet] Starting Trial 1 with timeout 2400.0 seconds

[integration_py2_mxnet] All retry failed.

CI_test.Makefile:152: recipe for target 'integration_py2_mxnet' failed

make: *** [integration_py2_mxnet] Error 1
```

rkooo567 commented:
Jenkins test this please

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/2002/
Test FAILed.

Commit:
Merge branch 'develop' into keras_deployer

withsmilo commented:
Jenkins test this please

withsmilo commented:
@rkooo567 https://amplab.cs.berkeley.edu/jenkins/job/Clipper-PRB/2005/ shows,
```
 [integration_py2_keras] Traceback (most recent call last): 
 [integration_py2_keras] File "/clipper/integration-tests/deploy_keras_models.py", line 12, in <module> 
 [integration_py2_keras] import keras 
 [integration_py2_keras] ImportError: No module named keras 
```

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/2005/
Test FAILed.

rkooo567 committed:
Added keras to devDockerfiles

rkooo567 committed:
Added keras to devDockerfiles

rkooo567 commented:
Jenkins test this please

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/2008/
Test FAILed.

simon-mo committed:
Merge branch 'develop' into keras_deployer

simon-mo commented:
jenkins test this please

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/2011/
Test FAILed.

rkooo567 commented:
jenkins test this please

rkooo567 commented:
jenkins test this please

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/2014/
Test FAILed.

rkooo567 committed:
Make model path abstract

rkooo567 committed:
Merge branch 'keras_deployer' of https://github.com/eavidan/clipper into keras_deployer

rkooo567 commented:
jenkins test this please

rkooo567 commented:
jenkins test this please

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/2020/
Test FAILed.

rkooo567 committed:
remove mxnet test to pass keras test. mxnet is very unstable now.

rkooo567 commented:
jenkins test this please

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/2022/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/2024/
Test FAILed.

rkooo567 commented:
@simon-mo So, there are 2 things.

1. I got this error log.
```
BuildError: pull access denied for keras-container, repository does not exist or may require 'docker login'
```

Do I need any additional setup for pulling an image?
EDIT -> I think we should create a repo at https://hub.docker.com/u/clipper/?page=2

2. Keras log is not organized at http://clipper-jenkins-viewer.herokuapp.com/. Do you need any steps to take to enable it?

Also, it looks like http://clipper-jenkins-viewer.herokuapp.com/ is broken for now?

rkooo567 committed:
Changed the image name to default

rkooo567 commented:
jenkins ok to test

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/2034/
Test PASSed.

simon-mo commented on bin/shipyard/clipper_test.cfg.py:
```diff
@@ -15,16 +15,17 @@
     "pysparkml": "python /clipper/integration-tests/deploy_pyspark_sparkml_models.py",
     "tensorflow": "python /clipper/integration-tests/deploy_tensorflow_models.py",
-    "mxnet": "python /clipper/integration-tests/deploy_mxnet_models.py",
+    #"mxnet": "python /clipper/integration-tests/deploy_mxnet_models.py",
```
revert this

rkooo567 replied:
reverted

simon-mo reviewed (changes requested):
lgtm

withsmilo commented on dockerfiles/ClipperDevDockerfile:
```diff
@@ -20,7 +20,7 @@ RUN pip install --upgrade pip
 RUN pip install -q cloudpickle==0.5.* pyzmq==17.0.* requests==2.20.0 subprocess32==3.2.* scikit-learn==0.19.* \
   numpy==1.14.* pyyaml>=4.2b1 docker==3.1.* kubernetes==6.0.* tensorflow==1.6.* mxnet==1.4.* pyspark==2.3.* \
-  xgboost==0.7.* jsonschema==2.6.* psutil==5.4.* prometheus_client
+  xgboost==0.7.* jsonschema==2.6.* psutil==5.4.* prometheus_client keras
```
How about set Keras' version like this, `keras==2.2.*`? See https://pypi.org/project/Keras/#history.

rkooo567 replied:
I agree. I will fix it. Thanks for the feedback!

withsmilo commented on dockerfiles/ClipperPy35DevDockerfile:
```diff
@@ -25,8 +25,7 @@ RUN echo '#!/bin/bash\npython3 "$@"' > /usr/bin/python && \
   numpy==1.14.* pyyaml>=4.2b1 docker==3.1.* kubernetes==5.0.* tensorflow==1.6.* mxnet==1.4.* pyspark==2.3.* \
-  xgboost==0.7.* urllib3==1.24.* # CI is broken when urllib3's version is 1.25.1. Delete urllib3==1.24.* later once version compatibility is stabilized
-
+  xgboost==0.7.* urllib3==1.24.* keras # CI is broken when urllib3's version is 1.25.1. Delete urllib3==1.24.* later once version compatibility is stabilized
```
How about set Keras' version like this, `keras==2.2.*`? See https://pypi.org/project/Keras/#history.

withsmilo commented on dockerfiles/KerasContainerDockerfile:
```diff
@@ -0,0 +1,13 @@
+ARG RPC_VERSION
+FROM ${REGISTRY}/${RPC_VERSION}-rpc:${CODE_VERSION}
+
+RUN pip install -q keras
```
How about set Keras' version like this, `keras==2.2.*`? See https://pypi.org/project/Keras/#history.

rkooo567 committed:
Code review fix

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/2036/
Test PASSed.

rkooo567 commented:
@simon-mo @withsmilo I think it is ready to go. @simon-mo Please add a keras-container repo at dockerhub, so that people can pull the image outside CI environment.

rkooo567 committed:
Merge branch 'develop' into keras_deployer

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/2037/
Test PASSed.

simon-mo commented:
@rkooo567 the docker repo will be automatically created when we push to master. I restarted travis.
